apiserver/logtransfer.go

```diff
@@ -54,26 +56,63 @@ func (s *migrationLoggingStrategy) Authenticate(req *http.Request) error {
 func (s *migrationLoggingStrategy) Start() {
 	s.filePrefix = s.st.ModelUUID() + ":"
 	s.dbLogger = state.NewDbLogger(s.st)
+	s.tracker = newLogTracker(s.st)
 }
 
 // Log writes the given record to the DB and to the backup file
 // logger. Part of LoggingStrategy.
 func (s *migrationLoggingStrategy) Log(m params.LogRecord) bool {
 	level, _ := loggo.ParseLevel(m.Level)
 	dbErr := s.dbLogger.Log(m.Time, m.Entity, m.Module, m.Location, level, m.Message)
+	if dbErr == nil {
+		dbErr = s.tracker.Track(m.Time)
+	}
 	if dbErr != nil {
 		logger.Errorf("logging to DB failed: %v", dbErr)
 	}
+
 	fileErr := logToFile(s.fileLogger, s.filePrefix, m)
 	if fileErr != nil {
 		logger.Errorf("logging to file logger failed: %v", fileErr)
 	}
+
 	return dbErr == nil && fileErr == nil
 }
 
 // Stop imdicates that there are no more log records coming, so we can
 // release resources and close loggers. Part of LoggingStrategy.
 func (s *migrationLoggingStrategy) Stop() {
 	s.dbLogger.Close()
+	s.tracker.Close()
 	s.ctxt.release(s.st)
 }
+
+const trackingPeriod = 2 * time.Minute
+
+func newLogTracker(st *state.State) *logTracker {
+	return &logTracker{tracker: state.NewLastSentLogTracker(st, st.ModelUUID(), "migration-logtransfer")}
+}
+
+type logTracker struct {
```

Comment:
The tracker only works correctly as long as the logs are seen in order right?

I realise that this is probably the case but it might be worth documenting that around here.